thread/common/thrprof.c

```diff
@@ -1025,3 +1030,64 @@ omrthread_get_jvm_cpu_usage_info_error_recovery(void)
 		GLOBAL_UNLOCK_SIMPLE(lib);
 	}
 }
+
+intptr_t
+omrthread_get_thread_times(omrthread_thread_time_t *threadTime)
+{
+#if defined(LINUX)
+	struct rusage rUsage;
+	memset(&rUsage, 0, sizeof(rUsage));
+
+	if (0 == getrusage(RUSAGE_THREAD, &rUsage)) {
+		threadTime->userTime = (SEC_TO_NANO_CONVERSION_CONSTANT * (int64_t)rUsage.ru_utime.tv_sec)
+			+ (MICRO_TO_NANO_CONVERSION_CONSTANT * (int64_t)rUsage.ru_utime.tv_usec);
+		threadTime->sysTime = (SEC_TO_NANO_CONVERSION_CONSTANT * (int64_t)rUsage.ru_stime.tv_sec)
+			+ (MICRO_TO_NANO_CONVERSION_CONSTANT * (int64_t)rUsage.ru_stime.tv_usec);
+
+		return 0;
+	}
+
+	return -1;
+#elif defined(OMR_OS_WINDOWS) && !defined(BREW) /* defined(LINUX) */
+	omrthread_t self = omrthread_self();
+	FILETIME creationTime;
+	FILETIME exitTime;
+	FILETIME kernelTime;
+	FILETIME userTime;
+	memset(&creationTime, 0, sizeof(creationTime));
+	memset(&exitTime, 0, sizeof(exitTime));
+	memset(&kernelTime, 0, sizeof(kernelTime));
+	memset(&userTime, 0, sizeof(userTime));
+
+	if (GetThreadTimes(self->handle, &creationTime, &exitTime, &kernelTime, &userTime)) {
+		/* Time is in 100's of nanos. Convert to nanos. */
+		threadTime->sysTime = ((int64_t)kernelTime.dwLowDateTime | ((int64_t)kernelTime.dwHighDateTime << 32)) * 100;
+		threadTime->userTime = ((int64_t)userTime.dwLowDateTime | ((int64_t)userTime.dwHighDateTime << 32)) * 100;
```

Comment:
When `dwLowDateTime` is negative, this yields the wrong result (it ignores `dwHighDateTime`).

See advice at https://learn.microsoft.com/en-us/windows/win32/api/minwinbase/ns-minwinbase-filetime and example code in [`gtest.cc`](https://github.com/eclipse/omr/blob/master/third_party/gtest-1.8.0/src/gtest.cc#L821-L824).

Reply:
Can you explain how could `dwLowDateTime` be negative?

Reply:
Ok, `dwLowDateTime` uses the type `DWORD` which is equivalent to `uint32_t` (so it can't be negative). Either we need an extra cast to `uint64_t` before casting to `int64_t` (casts should not change both the width and signedness), or (preferred) we should follow the advice from Microsoft to use `ULARGE_INTEGER` and its `QuadPart` member like the example in ``gtest.cc` I referenced.

Reply:
I think Microsoft's advice is more about manipulating the time value so using `ULARGE_INTEGER` just to get the time value seems unnecessary to me. I'll open a PR to fix all the related casting.

Reply:
Opened a PR: https://github.com/eclipse/omr/pull/7507